#### HEALTH CARE ANALYTICS PROJECT

##### 1. Predict likelihood of having health insurance based on demographics.             
##### 2. Predict frequency of preventive care visits (e.g., routine check-ups).
##### 3. Identify factors influencing cancer screening participation.
##### 4. Analyze healthcare access patterns across different geographic areas.
##### 5. Predict time until the next medical visit.


### DATA LOADING

In [15]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import scipy.stats as stats


In [16]:
# openpyxl is a Python library used to read, write, and manipulate Excel files in the .xlsx format. It serves as a bridge for handling Excel data seamlessly within Python scripts.
# File path
file_path = 'Copy of Healthcare Dataset..xlsx'

# Storing the dataset in a dataframe
df = pd.read_excel(file_path)


In [17]:
# Reading the head of the dataframe
df.head()

Location  _Location_latitude  \
0   -0.2742007 36.058336 1882.2000732421875 20.0           -0.274201   
1  -0.7158125 37.1475058 1361.9000244140625 20.0           -0.715812   
2  -0.7158157 37.1475082 1361.9000244140625 20.0           -0.715816   
3    -0.7157082 37.14749 1361.9000244140625 20.0           -0.715708   
4  -0.7157337 37.1474799 1361.9000244140625 20.0           -0.715734   

   _Location_longitude  _Location_altitude  _Location_precision  \
0            36.058336         1882.200073                 20.0   
1            37.147506         1361.900024                 20.0   
2            37.147508         1361.900024                 20.0   
3            37.147490         1361.900024                 20.0   
4            37.147480         1361.900024                 20.0   

        Date and Time    Age  Gender Marital Status  \
0                 NaT  41-50  Female        Married   
1 2023-05-15 13:38:00  18-30    Male         Single   
2 2023-05-15 13:49:00  41-50  Female        Married   
3 2023-05-15 14:19:00  18-30    Male         Single   
4 2023-05-15 14:30:00  41-50    Male        Married   

   How many children do you have, if any?  ...        _id  \
0                                     2.0  ...  230162389   
1                                     0.0  ...  238299334   
2                                     5.0  ...  238301092   
3                                     NaN  ...  238311207   
4                                     7.0  ...  238314477   

                                  _uuid    _submission_time  \
0  aa30304f-84f2-4c1b-b30a-371241f2ff17 2023-04-05 08:44:06   
1  63c461e3-b3ef-47cf-9632-0c912a639f46 2023-05-15 10:44:01   
2  4209a55d-a983-433f-8ce0-bce6cd28d713 2023-05-15 10:51:08   
3  2eba9b13-1706-4faf-b7a7-e45e9dcf48ab 2023-05-15 11:22:19   
4  1f49cb82-75ff-4391-b69b-973a6ce698ee 2023-05-15 11:33:21   

  _validation_status  _notes            _status _submitted_by  \
0                NaN     NaN  submitted_via_web           NaN   
1                NaN     NaN  submitted_via_web    safra_data   
2                NaN     NaN  submitted_via_web    safra_data   
3                NaN     NaN  submitted_via_web    safra_data   
4                NaN     NaN  submitted_via_web    safra_data   

              __version__ _tags _index  
0  vJ8gEKnN2pccxThc5jnkz4   NaN      1  
1  vMrCPR7NLZZJrf4PTsQ8uH   NaN      2  
2  vMrCPR7NLZZJrf4PTsQ8uH   NaN      3  
3  vMrCPR7NLZZJrf4PTsQ8uH   NaN      4  
4  vMrCPR7NLZZJrf4PTsQ8uH   NaN      5  

[5 rows x 32 columns]

In [18]:
# Reading the tail
df.tail()

Location  _Location_latitude  \
6153              -1.2693104 36.8916948 1618.0 4.816           -1.269310   
6154    -1.2705219 36.8923876 1594.9000244140625 5.0           -1.270522   
6155   -1.2718084 36.8931014 1594.9000244140625 4.84           -1.271808   
6156   -1.2730717 36.8938328 1594.9000244140625 4.84           -1.273072   
6157  -1.2739374 36.8942481 1594.9000244140625 4.616           -1.273937   

      _Location_longitude  _Location_altitude  _Location_precision  \
6153            36.891695         1618.000000                4.816   
6154            36.892388         1594.900024                5.000   
6155            36.893101         1594.900024                4.840   
6156            36.893833         1594.900024                4.840   
6157            36.894248         1594.900024                4.616   

           Date and Time    Age  Gender Marital Status  \
6153 2023-07-19 08:55:00  18-30  Female         Single   
6154 2023-07-19 08:59:00  41-50    Male        Married   
6155 2023-07-19 09:56:00  31-40  Female        Married   
6156 2023-07-19 10:02:00  31-40    Male       Divorced   
6157 2023-07-19 11:05:00  31-40  Female        Married   

      How many children do you have, if any?  ...        _id  \
6153                                     NaN  ...  258479363   
6154                                     6.0  ...  258479365   
6155                                     3.0  ...  258479369   
6156                                     2.0  ...  258479374   
6157                                     1.0  ...  258479375   

                                     _uuid    _submission_time  \
6153  4634218f-df21-4e98-a159-6b87edc11b35 2023-08-07 09:12:11   
6154  42076e39-7c01-4f78-a894-2a2158d61798 2023-08-07 09:12:12   
6155  ee33d8cb-4b92-4f98-9ab2-b61ffd5d3a84 2023-08-07 09:12:12   
6156  66c4ac84-e315-4197-8d3b-d814b837f7e6 2023-08-07 09:12:13   
6157  a9afdc88-0163-4af5-9b5d-64c277e17f54 2023-08-07 09:12:14   

     _validation_status  _notes            _status _submitted_by  \
6153                NaN     NaN  submitted_via_web    safra_data   
6154                NaN     NaN  submitted_via_web    safra_data   
6155                NaN     NaN  submitted_via_web    safra_data   
6156                NaN     NaN  submitted_via_web    safra_data   
6157                NaN     NaN  submitted_via_web    safra_data   

                 __version__ _tags _index  
6153  vNRCdRrE8Lpm9kXT55RgPB   NaN   6154  
6154  vNRCdRrE8Lpm9kXT55RgPB   NaN   6155  
6155  vNRCdRrE8Lpm9kXT55RgPB   NaN   6156  
6156  vNRCdRrE8Lpm9kXT55RgPB   NaN   6157  
6157  vNRCdRrE8Lpm9kXT55RgPB   NaN   6158  

[5 rows x 32 columns]

In [19]:
# Generating the information of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6158 entries, 0 to 6157
Data columns (total 32 columns):
 #   Column                                                                                                                      Non-Null Count  Dtype         
---  ------                                                                                                                      --------------  -----         
 0   Location                                                                                                                    5805 non-null   object        
 1   _Location_latitude                                                                                                          5805 non-null   float64       
 2   _Location_longitude                                                                                                         5805 non-null   float64       
 3   _Location_altitude                                                                      

##### The dataset contains 6,158 rows and 32 columns, representing healthcare-related data collected from respondents. It includes a mix of categorical, numerical, and datetime data types. The columns provide information about demographics, healthcare access, and preventive care trends. Key features include location data (latitude, longitude, altitude) for 5,805 entries, and a timestamp for 6,010 entries. Demographic details like Age (6,140 entries), Gender (6,141), Marital Status (6,140), and Employment Status (6,134) are present, with some missing values.

##### Healthcare-related columns track insurance coverage, with "Have you ever had health insurance?" containing 6,139 non-null entries and "If yes, which insurance cover?" having only 3,618 non-null entries. Preventive care participation, such as routine check-ups and cancer screenings, has substantial missing data for the frequency columns, with only 1,776 and 1,565 non-null values, respectively. Numeric columns like "How many children do you have?" (5,533 non-null) and "Time since last hospital visit" (6,000 non-null) capture additional insights.

##### Several columns, such as "Your Picture," "Your Picture_URL," "validation status," and "_tags," contain no data (0 non-null). The dataset also includes metadata like "_id," "_uuid," "_submission_time," and "_status", which are fully populated. Overall, the dataset is large and detailed but contains missing values in key fields related to healthcare access and preventive care, requiring preprocessing and cleaning. The total memory usage is approximately 1.5 MB.

In [20]:
# Displaying all the column names of the dataset
from tabulate import tabulate
column_names = df.columns.tolist()
# Convert all the column names to a list
column_names_table = [[i+1, name] for i, name in enumerate(column_names)]
print(tabulate(column_names_table, headers=["Index", "Column Name"]))   

  Index  Column Name
-------  --------------------------------------------------------------------------------------------------------------------------
      1  Location
      2  _Location_latitude
      3  _Location_longitude
      4  _Location_altitude
      5  _Location_precision
      6  Date and Time
      7  Age
      8  Gender
      9  Marital Status
     10  How many children do you have, if any?
     11  Employment Status
     12  Monthly Household Income
     13  Have you ever had health insurance?
     14  If yes, which insurance cover?
     15  When was the last time you visited a hospital for medical treatment? (In Months)
     16  Did you have health insurance during your last hospital visit?
     17  Have you ever had a routine check-up with a doctor or healthcare provider?
     18  If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?
     19  Have you ever had a cancer screening (e.g. mammogram, colo

### DATA CLEANING

In [21]:
# List of columns to drop
columns_to_drop = [
    'Your Picture', 'Your Picture_URL', '_validation_status', '_notes', '_tags',
    '_id', '_uuid', '_submission_time', '__version__', '_index'
]

# Drop the columns from the DataFrame
healthcare = df.drop(columns=columns_to_drop, errors='ignore')

# Displaying all the column names of the dataset
from tabulate import tabulate
column_names = healthcare.columns.tolist()
# Convert all the column names to a list
column_names_table = [[i+1, name] for i, name in enumerate(column_names)]
print(tabulate(column_names_table, headers=["Index", "Column Name"]))   

  Index  Column Name
-------  --------------------------------------------------------------------------------------------------------------------------
      1  Location
      2  _Location_latitude
      3  _Location_longitude
      4  _Location_altitude
      5  _Location_precision
      6  Date and Time
      7  Age
      8  Gender
      9  Marital Status
     10  How many children do you have, if any?
     11  Employment Status
     12  Monthly Household Income
     13  Have you ever had health insurance?
     14  If yes, which insurance cover?
     15  When was the last time you visited a hospital for medical treatment? (In Months)
     16  Did you have health insurance during your last hospital visit?
     17  Have you ever had a routine check-up with a doctor or healthcare provider?
     18  If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?
     19  Have you ever had a cancer screening (e.g. mammogram, colo

In [22]:
import warnings
warnings.filterwarnings('ignore')

# Replace '4+' with 5 (representing 4 and above)
healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?'] = healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?'].replace('4+', 5)
healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?'] = healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?'].replace('4+', 5)    


In [10]:
# Separate numerical and non-numerical columns
numerical_columns = healthcare.select_dtypes(include=['number']).columns
non_numerical_columns = healthcare.select_dtypes(exclude=['number']).columns

print("Numerical Columns: ", numerical_columns)
print("Non-Numerical Columns: ", non_numerical_columns)


Numerical Columns:  Index(['_Location_latitude', '_Location_longitude', '_Location_altitude',
       '_Location_precision', 'How many children do you have, if any?',
       'When was the last time you visited a hospital for medical treatment? (In Months)',
       'If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?',
       'If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?'],
      dtype='object')
Non-Numerical Columns:  Index(['Location', 'Date and Time', 'Age', 'Gender', 'Marital Status',
       'Employment Status', 'Monthly Household Income',
       'Have you ever had health insurance?', 'If yes, which insurance cover?',
       'Did you have health insurance during your last hospital visit?',
       'Have you ever had a routine check-up with a doctor or healthcare provider?',
       'Have you ever had a cancer screening (e.g. mammogram, col

In [23]:
# Check for missing values in all columns
missing_values = healthcare.isnull().sum()
print(missing_values)

Location                                                                                                                       353
_Location_latitude                                                                                                             353
_Location_longitude                                                                                                            353
_Location_altitude                                                                                                             353
_Location_precision                                                                                                            353
Date and Time                                                                                                                  148
Age                                                                                                                             18
Gender                                                                             

In [24]:
# Drop rows with missing values in the specified columns
columns_to_check = [
    'Location', '_Location_latitude', '_Location_longitude', '_Location_altitude', '_Location_precision', 
    'Age', 'Gender', 'Marital Status', 'Have you ever had a routine check-up with a doctor or healthcare provider?', 
    'Have you ever had a cancer screening (e.g. mammogram, colonoscopy, etc.)?','Have you ever had health insurance?',
    'Did you have health insurance during your last hospital visit?','_submitted_by'
]

healthcare = healthcare.dropna(subset=columns_to_check)

# Verify the changes
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5684 entries, 1 to 6157
Data columns (total 22 columns):
 #   Column                                                                                                                      Non-Null Count  Dtype         
---  ------                                                                                                                      --------------  -----         
 0   Location                                                                                                                    5684 non-null   object        
 1   _Location_latitude                                                                                                          5684 non-null   float64       
 2   _Location_longitude                                                                                                         5684 non-null   float64       
 3   _Location_altitude                                                                           

In [25]:
import warnings

# Ignore warnings
warnings.filterwarnings('ignore')

# Fill missing values with zero
healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?'].fillna(0, inplace=True)
healthcare['If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?'].fillna(0, inplace=True)
healthcare['If yes, which insurance cover?'].fillna(0, inplace=True)
healthcare['How many children do you have, if any?'].fillna(0, inplace=True)


In [26]:
# Fill missing values in 'Employment Status' with the mode
employment_mode = healthcare['Employment Status'].mode()[0]
healthcare['Employment Status'].fillna(employment_mode, inplace=True)

# Fill missing values in 'Monthly Household Income' with the mode
income_mode = healthcare['Monthly Household Income'].mode()[0]
healthcare['Monthly Household Income'].fillna(income_mode, inplace=True)

# Fill the missing values in 'When was the last time you visited a hospital for medical treatment? (In Months)' with the mode
last_visit_mode = healthcare['When was the last time you visited a hospital for medical treatment? (In Months)'].mode()[0]
healthcare['When was the last time you visited a hospital for medical treatment? (In Months)'].fillna(last_visit_mode, inplace=True)

# Fill the missing values in 'Date and Time' with the mode
date_mode = healthcare['Date and Time'].mode()[0]
healthcare['Date and Time'].fillna(date_mode, inplace=True)

In [27]:
# Recheking for missing values
missing_values = healthcare.isnull().sum()

print(missing_values)

Location                                                                                                                      0
_Location_latitude                                                                                                            0
_Location_longitude                                                                                                           0
_Location_altitude                                                                                                            0
_Location_precision                                                                                                           0
Date and Time                                                                                                                 0
Age                                                                                                                           0
Gender                                                                                                  

In [28]:
# Checking for duplicates
healthcare.duplicated().sum()

np.int64(0)

In [29]:
# Function to detect outliers using the IQR method
def detect_outliers_iqr(healthcare):
    outlier_counts = {}
    # Only consider numeric columns
    for column in healthcare.select_dtypes(include=['float', 'int']).columns:  
        Q1 = healthcare[column].quantile(0.25)  # 1st Quartile
        Q3 = healthcare[column].quantile(0.75)  # 3rd Quartile
        IQR = Q3 - Q1                          # Interquartile Range
        
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        
        # Count outliers for the current column
        outliers = healthcare[(healthcare[column] < lower_bound) | (healthcare[column] > upper_bound)]
        outlier_counts[column] = len(outliers)
    
    return pd.Series(outlier_counts, name='Number of Outliers')

# Apply the function to detect outliers and display the output
outliers = detect_outliers_iqr(healthcare)
print("Number of Outliers in Each Column:\n", outliers)


Number of Outliers in Each Column:
 _Location_latitude                                                                                                             583
_Location_longitude                                                                                                            361
_Location_altitude                                                                                                             256
_Location_precision                                                                                                           1199
How many children do you have, if any?                                                                                         236
When was the last time you visited a hospital for medical treatment? (In Months)                                               393
If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?     523
If you answered yes to the previous question, w

In [30]:
# Function to cap outliers using IQR method
def cap_outliers_iqr(healthcare):
    for column in healthcare.select_dtypes(include=['float', 'int']).columns:
        # Calculate IQR
        Q1 = healthcare[column].quantile(0.25)  # 1st Quartile
        Q3 = healthcare[column].quantile(0.75)  # 3rd Quartile
        IQR = Q3 - Q1                    # Interquartile Range
        
        # Define lower and upper bounds
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        
        # Cap the outliers
        healthcare[column] = healthcare[column].apply(lambda x: 
                                         lower_bound if x < lower_bound else
                                         upper_bound if x > upper_bound else x)
    return healthcare

# Apply the function to cap outliers
health = cap_outliers_iqr(healthcare)

# Verify changes
print("Outliers capped successfully.")


Outliers capped successfully.


In [31]:
# Ensure Date columns are datetime64
health['Date and Time'] = pd.to_datetime(health['Date and Time'])
# Verify the changes
print(health[['Date and Time']].dtypes)

Date and Time    datetime64[ns]
dtype: object


In [32]:
# Flag inconsistencies between health insurance responses
inconsistent_rows = health[
    (health['Have you ever had health insurance?'] == 'No') & 
    (health['Did you have health insurance during your last hospital visit?'] == 'Yes')
]

# Display inconsistent rows
print("Inconsistent Rows:")
print(inconsistent_rows)



Inconsistent Rows:
                                               Location  _Location_latitude  \
2         -0.7158157 37.1475082 1361.9000244140625 20.0           -0.715816   
12                    0.3366065 37.5641427 1140.7 4.957            0.336606   
109               -0.3035 36.085388333333334 1788.1 4.9           -0.303500   
112   -0.30364833333333335 36.08537666666667 1787.6 4.9           -0.303648   
113           -0.30361166666666667 36.085375 1787.7 5.0           -0.303612   
...                                                 ...                 ...   
5676                  0.8017472 36.6521847 1819.5 4.775            0.801747   
5681                    0.828125 36.697271 1761.4 4.816            0.828125   
5901                   1.8421569 36.8208419 1429.0 4.98            0.938998   
6023                   -0.970085 36.8429787 1856.8 3.75           -0.970085   
6047                   -0.8941817 36.8144287 0.0 3200.0           -0.894182   

      _Location_longitude  _Loca

##### Ensuring data Intergrity

In [34]:
# Checking for Data Integrity
# Cross-check logical relationships between columns

# Flag inconsistencies between health insurance responses
inconsistent_rows = health[
    (health['Have you ever had health insurance?'] == 'No') & 
    (health['Did you have health insurance during your last hospital visit?'] == 'Yes')
]

# Display inconsistent rows
print("Inconsistent Rows:")
print(inconsistent_rows)

# Correct inconsistencies by setting 'Did you have health insurance during your last hospital visit?' to 'No'
health.loc[inconsistent_rows.index, 'Did you have health insurance during your last hospital visit?'] = 'No'

# Verify corrections
corrected_inconsistent_rows = health[
    (health['Have you ever had health insurance?'] == 'No') & 
    (health['Did you have health insurance during your last hospital visit?'] == 'Yes')
]

print("Corrected Inconsistent Rows:")
print(corrected_inconsistent_rows)

Inconsistent Rows:
Empty DataFrame
Columns: [Location, _Location_latitude, _Location_longitude, _Location_altitude, _Location_precision, Date and Time, Age, Gender, Marital Status, How many children do you have, if any?, Employment Status, Monthly Household Income, Have you ever had health insurance?, If yes, which insurance cover?, When was the last time you visited a hospital for medical treatment? (In Months), Did you have health insurance during your last hospital visit?, Have you ever had a routine check-up with a doctor or healthcare provider?, If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?, Have you ever had a cancer screening (e.g. mammogram, colonoscopy, etc.)?, If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?, _status, _submitted_by]
Index: []

[0 rows x 22 columns]
Corrected Inconsistent Rows:
Empty DataFrame
Columns: [Location

In [35]:
health.shape

(5684, 22)

In [36]:
# Convert all columns to strings
health[non_numerical_columns] = health[non_numerical_columns].astype(str)

# Standardizing text data for non-numerical columns
def standardize_text_data(df, columns):
    for column in columns:
        df[column] = df[column].str.lower().str.strip()
    return df

# Apply the function to standardize text data
health = standardize_text_data(health, non_numerical_columns)

# Verify the changes
print(health[non_numerical_columns].head())

                                          Location        Date and Time  \
1    -0.7158125 37.1475058 1361.9000244140625 20.0  2023-05-15 13:38:00   
2    -0.7158157 37.1475082 1361.9000244140625 20.0  2023-05-15 13:49:00   
3      -0.7157082 37.14749 1361.9000244140625 20.0  2023-05-15 14:19:00   
4    -0.7157337 37.1474799 1361.9000244140625 20.0  2023-05-15 14:30:00   
5  -0.7158041 37.1475364 1361.9000244140625 26.107  2023-05-15 14:35:00   

     Age  Gender Marital Status Employment Status Monthly Household Income  \
1  18-30    male         single        unemployed          less than 10000   
2  41-50  female        married     self-employed              20001-30000   
3  18-30    male         single     self-employed              10001-20000   
4  41-50    male        married     self-employed              20001-30000   
5  18-30  female         single        unemployed              10001-20000   

  Have you ever had health insurance? If yes, which insurance cover?  \
1       

#### DATA PRE-PROCESSING FOR MACHINE LEARNING

#### EXPLANATORY DATA ANALYSIS

In [37]:
# Getting the description of the dataset
healthcare.describe()

_Location_latitude  _Location_longitude  _Location_altitude  \
count         5684.000000          5684.000000         5684.000000   
mean            -0.754034            36.760739         1559.674100   
std              0.787198             0.731454          430.663920   
min             -2.571207            35.218006          583.705222   
25%             -1.254880            36.375370         1347.975006   
50%             -0.724745            36.857319         1592.900024   
75%             -0.377329            37.146946         1857.488196   
max              0.938998            38.304310         2621.757980   

       _Location_precision  How many children do you have, if any?  \
count          5684.000000                             5684.000000   
mean              5.884590                                2.233110   
std               2.989164                                1.852094   
min               0.000000                                0.000000   
25%               4.100000                                1.000000   
50%               4.820000                                2.000000   
75%               6.900000                                3.000000   
max              11.100000                                6.000000   

       When was the last time you visited a hospital for medical treatment? (In Months)  \
count                                        5684.000000                                  
mean                                            5.282996                                  
std                                             3.991392                                  
min                                             0.000000                                  
25%                                             2.000000                                  
50%                                             4.000000                                  
75%                                             7.000000                                  
max                                            14.500000                                  

       If you answered yes to the previous question, what time period (in years) do you stay before having your routine check-up?  \
count                                        5684.000000                                                                            
mean                                            0.498505                                                                            
std                                             0.854364                                                                            
min                                             0.000000                                                                            
25%                                             0.000000                                                                            
50%                                             0.000000                                                                            
75%                                             1.000000                                                                            
max                                             2.500000                                                                            

       If you answered yes to the previous question, what time period (in years) do you stay before having your Cancer screening?  
count                                        5684.000000                                                                           
mean                                            0.512667                                                                           
std                                             0.923627                                                                           
min                                             0.000000                                                                           
25%                                             0.000000           